src/hotspot/share/prims/jvmtiEnvBase.cpp

```diff
@@ -1517,24 +1536,133 @@ JvmtiModuleClosure::get_all_modules(JvmtiEnv* env, jint* module_count_ptr, jobje
+  for (vframeStream vfs(java_thread, true, false /* process_frames */); !vfs.at_end(); vfs.next()) {
+    methodHandle mh(current_thread, vfs.method());
+    if (mh->is_native()) {
+      _result = JVMTI_ERROR_OPAQUE_FRAME;
+      return;
+    }
+    is_interpreted[frame_count] = vfs.is_interpreted_frame();
+    frame_sp[frame_count] = vfs.frame_id();
+    if (++frame_count > 1) break;
+  }
+  if (frame_count < 2)  {
+    // We haven't found two adjacent non-native Java frames on the top.
+    // There can be two situations here:
+    //  1. There are no more java frames
+    //  2. Two top java frames are separated by non-java native frames
+    if(JvmtiEnvBase::vframeForNoProcess(java_thread, 1) == NULL) {
+      _result = JVMTI_ERROR_NO_MORE_FRAMES;
+      return;
+    } else {
+      // Intervening non-java native or VM frames separate java frames.
+      // Current implementation does not support this. See bug #5031735.
+      // In theory it is possible to pop frames in such cases.
+      _result = JVMTI_ERROR_OPAQUE_FRAME;
+      return;
+    }
+  }
+
+  // If any of the top 2 frames is a compiled one, need to deoptimize it
+  for (int i = 0; i < 2; i++) {
+    if (!is_interpreted[i]) {
+      Deoptimization::deoptimize_frame(java_thread, frame_sp[i]);
+    }
+  }
+
+  // Update the thread state to reflect that the top frame is popped
+  // so that cur_stack_depth is maintained properly and all frameIDs
+  // are invalidated.
+  // The current frame will be popped later when the suspended thread
+  // is resumed and right before returning from VM to Java.
+  // (see call_VM_base() in assembler_<cpu>.cpp).
+
+  // It's fine to update the thread state here because no JVMTI events
+  // shall be posted for this PopFrame.
+
+  // It is only safe to perform the direct operation on the current
+  // thread. All other usage needs to use a handshake for safety.
   JavaThread* jt = _state->get_thread();
   assert(jt == target, "just checking");
   if (!jt->is_exiting() && jt->threadObj() != NULL) {
     _state->update_for_pop_top_frame();
+    java_thread->set_popframe_condition(JavaThread::popframe_pending_bit);
+    // Set pending step flag for this popframe and it is cleared when next
+    // step event is posted.
+    _state->set_pending_step_for_popframe();
     _result = JVMTI_ERROR_NONE;
   }
 }
 
 void
-SetFramePopClosure::do_thread(Thread *target) {
-  JavaThread* jt = _state->get_thread();
-  assert(jt == target, "just checking");
-  if (!jt->is_exiting() && jt->threadObj() != NULL) {
-    int frame_number = _state->count_frames() - _depth;
-    _state->env_thread_state((JvmtiEnvBase*)_env)->set_frame_pop(frame_number);
-    _result = JVMTI_ERROR_NONE;
+SetFramePopClosure::doit(Thread *target, bool self) {
+  ResourceMark rm;
+  JavaThread* java_thread = target->as_Java_thread();
+
+  assert(_state->get_thread() == java_thread, "Must be");
+
+  if (!self) {
+    if (!java_thread->is_external_suspend()) {
```

Comment:
You could join these two if-statements with `&&` and have
one less indenting level...